x-pack/metricbeat/module/aws/rds/rds.go

```diff
@@ -0,0 +1,223 @@
+	// Get startTime and endTime
+	startTime, endTime, err := aws.GetStartTimeEndTime(m.DurationString)
+	if err != nil {
+		m.logger.Error(errors.Wrap(err, "Error ParseDuration"))
+		report.Error(err)
+		return
+	}
+
+	for _, regionName := range m.MetricSet.RegionsList {
+		m.MetricSet.AwsConfig.Region = regionName
+		svc := rds.New(*m.MetricSet.AwsConfig)
+
+		describeInstanceInput := &rds.DescribeDBInstancesInput{}
+		req := svc.DescribeDBInstancesRequest(describeInstanceInput)
+		output, err := req.Send()
+		if err != nil {
+			err = errors.Wrap(err, "DescribeDBInstancesRequest failed, skipping region "+regionName)
+			m.logger.Errorf(err.Error())
+			report.Error(err)
+			continue
+		}
+		if len(output.DBInstances) == 0 {
+			continue
+		}
+		// get DBInstance ARN
+		dbInstanceArns := []string{}
+		for _, dbInstance := range output.DBInstances {
+			dbInstanceArns = append(dbInstanceArns, *dbInstance.DBInstanceArn)
+		}
+
+		svcCloudwatch := cloudwatch.New(*m.MetricSet.AwsConfig)
+		namespace := "AWS/RDS"
+		listMetricsOutput, err := aws.GetListMetricsOutput(namespace, regionName, svcCloudwatch)
+		if err != nil {
+			m.logger.Error(err.Error())
+			report.Error(err)
+			continue
+		}
+
+		if listMetricsOutput == nil || len(listMetricsOutput) == 0 {
+			continue
+		}
+
+		for _, dbInstanceArn := range dbInstanceArns {
+			metricDataQueries := constructMetricQueries(listMetricsOutput, dbInstanceArn, m.PeriodInSec)
+			// If metricDataQueries, still needs to createCloudWatchEvents.
+			metricDataOutput := []cloudwatch.MetricDataResult{}
+			if len(metricDataQueries) != 0 {
+				// Use metricDataQueries to make GetMetricData API calls
+				metricDataOutput, err = aws.GetMetricDataResults(metricDataQueries, svcCloudwatch, startTime, endTime)
+				if err != nil {
+					err = errors.Wrap(err, "GetMetricDataResults failed, skipping region "+regionName)
+					m.logger.Error(err.Error())
+					report.Error(err)
+					continue
+				}
+			}
+			// Create Cloudwatch Events for RDS
+			event, info, err := createCloudWatchEvents(metricDataOutput, dbInstanceArn, regionName)
+			if info != "" {
+				m.logger.Info(info)
+			}
+
+			if err != nil {
+				m.logger.Error(err.Error())
+				report.Error(err)
+				continue
+			}
+			report.Event(event)
+		}
+	}
+}
+
+func constructMetricQueries(listMetricsOutput []cloudwatch.Metric, dbInstanceArn string, periodInSec int) []cloudwatch.MetricDataQuery {
+	metricDataQueries := []cloudwatch.MetricDataQuery{}
+	metricDataQueryEmpty := cloudwatch.MetricDataQuery{}
+	for i, listMetric := range listMetricsOutput {
+		metricDataQuery := createMetricDataQuery(listMetric, i, dbInstanceArn, periodInSec)
+		if metricDataQuery == metricDataQueryEmpty {
+			continue
+		}
+		metricDataQueries = append(metricDataQueries, metricDataQuery)
+	}
+	return metricDataQueries
+}
+
+func createMetricDataQuery(metric cloudwatch.Metric, index int, dbInstanceArn string, periodInSec int) (metricDataQuery cloudwatch.MetricDataQuery) {
```

Comment:
nit: Is there any benefit to using named returns here? while the code is correct it usually makes it more difficult to read and understand what's the outcome of this function.

Reply:
Yeah good point, I will change in both places. Thank you!